### PR TITLE
[BYOC] Use Non-Recursive Visitor/Mutator

### PR DESCRIPTION
**Pull request by comaniac**
We have observed that recursive passes in BYOC infra will encounter the stack overflow issue when processing large model (e.g., ResNet-101). This PR refactors all but `PartitionGraph` passes in BYOC infra with non-recursive manner.

Another PR for `PartitionGraph` will be followed up. That pass does not traverse graph neither in pre-order nor in post-order, so we need more time to refactor.

cc @zhiics @mbrookhart @mbaret @manupa-arm

### CODE REVIEW

File: src/relay/transforms/annotate_target.cc [hunk 1]
```diff
@@ -181,50 +181,47 @@ class AnnotateTargetWrapper : public ExprMutator {
...
-    call->checked_type_ = cn->checked_type_;
+    Call new_call = Call(post_call->op, compiler_begins, post_call->attrs);
+    new_call->checked_type_ = pre->checked_type_;
 
     // Update the target map.
-    op_expr_to_target_[call] = target;
+    op_expr_to_target_[new_call] = target;
 
-    return std::move(call);
+    return new_call;
```
Comment (mbrookhart) [inline src/relay/transforms/annotate_target.cc:195]: I think this needs std::move still?

Comment (comaniac) [reply src/relay/transforms/annotate_target.cc:195]: Thanks for pointing out. I'll fix this along with other unit tests. Although it looks weird to me, I don't think that's a bug in the MixedMode classes.

**Comment by comaniac** [comment]
@mbrookhart I've fixed the unit tests which are related to the `AnnotateRegionSet`. Since `AnnotateRegionSet` was implemented using pre-order traversal (i.e., `process this node -> visit(node->args)`), I refactored its logic to post-order to fit to `MixedModeVisitor`. While passing all unit tests should guarantee its functionality, I'd be great if you could help checking if I miss-used `VisitExpr` in those passes. Thanks!

**Comment by zhiics** [comment]
@mbrookhart @comaniac The problem is not just about current pass that will be invoked, i.e. `AnnotateTarget`. We will update the module in the pass manager for function passes here:

https://github.com/apache/incubator-tvm/blob/f5c9bc93883d426284057657463942dfdfef2fd8/src/relay/ir/transform.cc#L140-L142

This will indirectly call InferType here:

https://github.com/apache/incubator-tvm/blob/f5c9bc93883d426284057657463942dfdfef2fd8/src/ir/module.cc#L225

Therefore, I think we may need to change `InferType` to non-recursive version first. Otherwise, many passes will still fail (indirectly/magically).

**Comment by zhiics** [comment]
Thanks @comaniac @mbrookhart @Somisary

**Pull request merged**